src/Draco.Compiler/Internal/Syntax/Formatting/DracoFormatter.cs

```diff
@@ -0,0 +1,440 @@
+        TokenKind.CurlyOpen => WhitespaceBehavior.SpaceBefore | WhitespaceBehavior.BehaveAsWhiteSpaceForNextToken,
+        TokenKind.ParenOpen => WhitespaceBehavior.BehaveAsWhiteSpaceForNextToken,
+        TokenKind.ParenClose => WhitespaceBehavior.BehaveAsWhiteSpaceForPreviousToken,
+        TokenKind.InterpolationStart => WhitespaceBehavior.Whitespace,
+        TokenKind.Dot => WhitespaceBehavior.Whitespace,
+        TokenKind.Colon => WhitespaceBehavior.BehaveAsWhiteSpaceForPreviousToken,
+
+        TokenKind.Assign => WhitespaceBehavior.PadAround,
+        TokenKind.LineStringStart => WhitespaceBehavior.SpaceBefore,
+        TokenKind.MultiLineStringStart => WhitespaceBehavior.SpaceBefore,
+        TokenKind.Plus => WhitespaceBehavior.SpaceBefore,
+        TokenKind.Minus => WhitespaceBehavior.SpaceBefore,
+        TokenKind.Star => WhitespaceBehavior.SpaceBefore,
+        TokenKind.Slash => WhitespaceBehavior.SpaceBefore,
+        TokenKind.PlusAssign => WhitespaceBehavior.SpaceBefore,
+        TokenKind.MinusAssign => WhitespaceBehavior.SpaceBefore,
+        TokenKind.StarAssign => WhitespaceBehavior.SpaceBefore,
+        TokenKind.SlashAssign => WhitespaceBehavior.SpaceBefore,
+        TokenKind.GreaterEqual => WhitespaceBehavior.SpaceBefore,
+        TokenKind.GreaterThan => WhitespaceBehavior.SpaceBefore,
+        TokenKind.LessEqual => WhitespaceBehavior.SpaceBefore,
+        TokenKind.LessThan => WhitespaceBehavior.SpaceBefore,
+        TokenKind.Equal => WhitespaceBehavior.SpaceBefore,
+        TokenKind.LiteralFloat => WhitespaceBehavior.SpaceBefore,
+        TokenKind.LiteralInteger => WhitespaceBehavior.SpaceBefore,
+
+        TokenKind.Identifier => WhitespaceBehavior.SpaceBefore,
+
+        _ => WhitespaceBehavior.NoFormatting
+    };
+
+    public override void VisitSyntaxToken(Api.Syntax.SyntaxToken node)
+    {
+        this.HandleTokenComments(node);
+        var formattingKind = GetFormattingTokenKind(node);
+
+        var notFirstToken = this.formatter.CurrentIdx > 0;
+        var doesntInsertSpace = !formattingKind.HasFlag(WhitespaceBehavior.SpaceBefore);
+        var doesReturnLine = this.formatter.CurrentToken.DoesReturnLine;
+        var insertNewline = doesReturnLine is not null && doesReturnLine.IsCompleted && doesReturnLine.Value;
+        var notWhitespaceNode = node.Kind != TokenKind.StringNewline && node.Kind != TokenKind.EndOfInput;
+        if (doesntInsertSpace && notFirstToken && !insertNewline && notWhitespaceNode)
+        {
+            var lexer = new Lexer(SourceReader.From(this.formatter.PreviousToken.Text + node.Text), default);
+            lexer.Lex();
+            var secondToken = lexer.Lex();
+            if (secondToken.Kind == TokenKind.EndOfInput) // this means the 2 tokens merged in a single one, we want to avoid that.
+            {
+                this.formatter.CurrentToken.Kind = WhitespaceBehavior.SpaceBefore;
+            }
```

Comment:
You could factor this out into some SyntaxFacts utiity method to see if you can paste 2 tokens together without them merging or something

Reply:
This was a hacky method which is probably very slow and I thought we would eventually come up with something better.